envoy/upstream/outlier_detection.h

```diff
@@ -47,6 +49,104 @@ enum class Result {
   ExtOriginRequestSuccess // Request was completed successfully.
 };
 
+// Types of errors reported to outlier detectors.
+// Each type may have a different syntax and content.
+// TODO (cpakulski) - maybe use enums from protobufs.
+enum class ExtResultType {
+  TEST, // Used in unit tests.
+  HTTP_CODE,
+  LOCAL_ORIGIN,
+};
+
+/*
+ * Class carries result of a transaction with upstream entity
+ * or generated internally by Envoy.
+ * Different categories of results will be derived from that base class.
+ * Those categories of results are fed only into Outlier Detection extensions.
+ */
+class ExtResult {
+public:
+  ExtResult() = delete;
+  ExtResult(ExtResultType type) : type_(type) {}
+  virtual ExtResultType type() const { return type_; };
+  virtual ~ExtResult() = default;
+
+private:
+  const ExtResultType type_;
+};
+
+// Types derived from TypedExtResult are used to report the result of transaction to
+// an outlier detection monitor.
+template <ExtResultType E> class TypedExtResult : public ExtResult {
+protected:
+  TypedExtResult() : ExtResult(E) {}
+};
+
+class HttpCode : public TypedExtResult<ExtResultType::HTTP_CODE> {
+public:
+  HttpCode(uint32_t code) : code_(code) {}
+  HttpCode() = delete;
+  virtual ~HttpCode() {}
+  uint32_t code() const { return code_; }
+
+private:
+  uint32_t code_;
+};
+
+// LocalOriginEvent is used to report errors like resets, timeouts but also
+// successful connection attempts.
+class LocalOriginEvent : public TypedExtResult<ExtResultType::LOCAL_ORIGIN> {
+public:
+  LocalOriginEvent(Result result) : result_(result) {}
+  LocalOriginEvent() = delete;
+  Result result() const { return result_; }
+
+private:
+  Result result_;
+};
+
+// Base class for various types of monitors.
+// Each monitor may implement different health detection algorithm.
+class ExtMonitor {
+public:
+  ExtMonitor(const std::string& name, uint32_t enforce) : name_(name), enforce_(enforce) {}
+  ExtMonitor() = delete;
+  virtual ~ExtMonitor() {}
+  virtual void reportResult(const ExtResult&) PURE;
+
+  void
+  setCallback(std::function<void(uint32_t, std::string, absl::optional<std::string>)> callback) {
+    callback_ = callback;
+  }
+
+  void reset() { onReset(); }
+  std::string name() const { return name_; }
+
+protected:
+  virtual bool onError() PURE;
+  virtual void onSuccess() PURE;
+  virtual void onReset() PURE;
+  virtual std::string getFailedExtraInfo() { return ""; }
+
+  std::string name_;
+  uint32_t enforce_{100};
+  std::function<void(uint32_t, std::string, absl::optional<std::string>)> callback_;
+};
+
+using ExtMonitorPtr = std::unique_ptr<ExtMonitor>;
+
+class ExtMonitorsSet {
+public:
+  void addMonitor(ExtMonitorPtr&& monitor) { monitors_.push_back(std::move(monitor)); }
+  void for_each(std::function<void(ExtMonitorPtr&)> f) {
+    for (auto& monitor : monitors_) {
+      f(monitor);
+    }
+  }
```

Comment:
```suggestion
  template<class CALLABLE> void forEach(const CALLABLE& f) {
    for (auto& monitor : monitors_) {
      f(*monitor);
    }
  }
```

Reply:
Great idea. Converted to template. Thanks.